pkg/controller/kubelet-config/kubelet_config_bootstrap.go

```diff
@@ -0,0 +1,166 @@
+package kubeletconfig
+
+import (
+	"encoding/json"
+	"fmt"
+
+	ign3types "github.com/coreos/ignition/v2/config/v3_2/types"
+	"github.com/imdario/mergo"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	mtmpl "github.com/openshift/machine-config-operator/pkg/controller/template"
+	"github.com/openshift/machine-config-operator/pkg/version"
+	"github.com/vincent-petithory/dataurl"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+)
+
+// RunKubeletBootstrap generates MachineConfig objects for mcpPools that would have been generated by syncKubeletConfig
+// nolint:gocyclo
+func RunKubeletBootstrap(templateDir string, kubeletConfig *mcfgv1.KubeletConfig, controllerConfig *mcfgv1.ControllerConfig, mcpPools []*mcfgv1.MachineConfigPool) ([]*mcfgv1.MachineConfig, error) {
+	var (
+		kubeletIgnition            *ign3types.File
+		logLevelIgnition           *ign3types.File
+		autoSizingReservedIgnition *ign3types.File
+	)
+	userDefinedSystemReserved := make(map[string]string)
+
+	// Validate the KubeletConfig CR if exists
+	if err := validateUserKubeletConfig(kubeletConfig); err != nil {
+		return nil, newForgetError(err)
+	}
+
+	var res []*mcfgv1.MachineConfig
+	for _, pool := range mcpPools {
```

Comment:
Edit: I just realized that we do have this default label:

https://github.com/openshift/machine-config-operator/blob/master/manifests/master.machineconfigpool.yaml#L8

Maybe for day 1 machineconfigs, we require this label to be specified for it to attach to the correct pool, and if otherwise not specified, we skip parsing that one for day 1?

Alternatively we can dynamically attempt to match pool labelling based on supplied labelling in the kubeletconfig here, since the MCP object should have already been parsed during bootstrap at this point

What do you think @QiWang19

Reply:
Hi @yuqi-zhang , updated with the pre-check of the default label (worker and master). Could you elaborate what are the possible concerns if the kubeletconfig is applied to pools without this check?

Reply:
Hi,

By default today, kubeletconfigs match based on pool labelling. You could have different kubeletconfigs per pool, as an example of this. You also don't necessarily want a kubeletconfig you only intend to apply to masters to apply to workers.

This will also immediately drift based on install time/post-install differences. The in-cluster MCO will generate kubeletconfig MCs based on pool labelling, so you will likely also error with rendered-xx not found, since they will generate different configs.

We also generally want to have install time behaviour match in-cluster behaviour. So even if the above error didn't apply, we still would want bootstrap manifest behaviour to match in-cluster manifest behaviour.

Hope that helps, let me know if you have any further questions